File: R/switch.R
#' Shiny Server for Geyser Data Switch
#'
#' @param id shiny identifier

#' @return reactive server
#' @export
#' @rdname switchServer
#' @importFrom shiny moduleServer NS renderUI selectInput shinyApp uiOutput
#' @importFrom bslib layout_columns page
switchServer <- function(id) {
  shiny::moduleServer(id, function(input, output, session) {
    ns <- session$ns
    
    # Module to select `dataset()`.
    dataset <- datasetsServer("datasets")
    # Modules to plot data.
    histServer("hist", dataset)
    gghistServer("gghist", dataset)
    ggpointServer("ggpoint", dataset)
    # Switches
    output$inputSwitch <- shiny::renderUI({
      shiny::req(input$plottype)
      get(paste0(input$plottype, "Input"))(ns(input$plottype))
    })
    output$uiSwitch <- shiny::renderUI({
      shiny::req(input$plottype)
      get(paste0(input$plottype, "UI"))(ns(input$plottype))
    })
    output$outputSwitch <- shiny::renderUI({
      shiny::req(input$plottype)
      get(paste0(input$plottype, "Output"))(ns(input$plottype))
    })
  })
}
#' Shiny Module Input
#' @return nothing returned
#' @rdname switchServer
#' @export
switchInput <- function(id) {
  ns <- shiny::NS(id)
  list(
    bslib::layout_columns(
      shiny::selectInput(ns("plottype"), "Plot Type:",
                         c("hist","gghist","ggpoint")),
      datasetsInput(ns("datasets")),
      datasetsUI(ns("datasets"))),
  shiny::uiOutput(ns("inputSwitch"))
  )
}
#' Shiny Module UI
#' @return nothing returned
#' @rdname switchServer
#' @export
switchUI <- function(id) {
  ns <- shiny::NS(id)
  shiny::uiOutput(ns("uiSwitch"))
}
#' Shiny Module Output
#' @return nothing returned
#' @rdname switchServer
#' @export
switchOutput <- function(id) {
  ns <- shiny::NS(id)
  shiny::uiOutput(ns("outputSwitch"))
}
#' Shiny Module App
#' @return nothing returned
#' @rdname switchServer
#' @export
switchApp <- function() {
  ui <- bslib::page(
    switchInput("switch"),
    switchOutput("switch"), 
    switchUI("switch")
  )
  server <- function(input, output, session) {
    switchServer("switch")
  }
  shiny::shinyApp(ui, server)
}

File: geyser/switch.py
from shiny import App, module, reactive, render, ui
import geyser.io as io
from geyser.hist import *
from geyser.gghist import *
from geyser.ggpoint import *
from geyser.datasets import *

module_dict = {
    "hist_input":     hist_input,
    "hist_output":    hist_output,
    "hist_ui":        hist_ui,
    "gghist_input":   gghist_input,
    "gghist_output":  gghist_output,
    "gghist_ui":      gghist_ui,
    "ggpoint_input":  ggpoint_input,
    "ggpoint_output": ggpoint_output,
    "ggpoint_ui":     ggpoint_ui
}

@module.server
def switch_server(input, output, session):
    """switch Server."""
    dataset = datasets_server("datasets")
    hist_server("hist", dataset)
    gghist_server("gghist", dataset)
    ggpoint_server("ggpoint", dataset)
    
    # Alternative: use match expr: case expr:.
    @render.ui
    def input_switch():
      return module_dict[f"{input.plottype()}_input"](input.plottype())
    @render.ui
    def output_switch():
      return module_dict[f"{input.plottype()}_output"](input.plottype())
    @render.ui
    def ui_switch():
      return module_dict[f"{input.plottype()}_ui"](input.plottype())

# switch_server("switch")

@module.ui
def switch_input():
    """switch Input."""
    return ui.card(
      ui.row(
        ui.column(4, ui.input_select("plottype", "Plot Type:",
            ["hist","gghist","ggpoint"])),
        ui.column(4, datasets_input("datasets")),
        ui.column(4, datasets_ui("datasets"))
      ),
      ui.output_ui("input_switch")
    )

# switch_input("switch")

@module.ui
def switch_output():
    """Switch Output."""
    return ui.output_ui("output_switch")
@module.ui
def switch_ui():
    """switch UI."""
    return ui.output_ui("ui_switch")

# switch_ui("switch")

def switch_app():
    """switch App."""
    app_ui = ui.page_fluid(
        switch_input("switch"),
        switch_output("switch"),
        switch_ui("switch")
    )
    def app_server(input, output, session):
        switch_server("switch")

    app = App(app_ui, app_server)

    #if __name__ == '__main__':
    io.app_run(app)

# hist_app()
